modules/ding_paragraphs/DingParagraphsBPIHelper.php

```diff
@@ -0,0 +1,365 @@
+      ));
+
+      $lang = $entity->language;
+
+      // Get paragraphs assets indexed by original url.
+      $paragraphs_assets = [];
+      if (isset($syndicated_images['paragraphs_image'])) {
+        foreach ($syndicated_images['paragraphs_image'] as $asset) {
+          $paragraphs_assets[$asset['external']] = $asset;
+        }
+      }
+
+      // Replace file urls with references to syndicated files.
+      $this->traverse($bpi_data, function ($key, &$value, &$data) use ($paragraphs_assets) {
+        if (is_array($value) && isset($value[self::BPI_FILE_URL])) {
+          $url = $value[self::BPI_FILE_URL];
+          if (isset($paragraphs_assets[$url]['@managed_file'])) {
+            // Inject the file object (as an array).
+            $value = (array) $paragraphs_assets[$url]['@managed_file'];
+          }
+          else {
+            // No syndicated file found. Remove the data entry.
+            unset($data[$key]);
+          }
+        }
+      });
+
+      // Build the final list of paragraphs.
+      foreach ($paragraphs_fields as $paragraphs_field_name => $paragraphs_field) {
+        if (isset($bpi_data[$paragraphs_field_name]) && is_array($bpi_data[$paragraphs_field_name])) {
+          foreach ($bpi_data[$paragraphs_field_name] as $item) {
+            $paragraphs_items[] = new ParagraphsItemEntity($item);
+          }
+        }
+      }
+    }
+
+    if (empty($paragraphs_items)) {
+      // If we don't have any paragraphs items, we create a single paragraph
+      // and set its body to the full body content from the BPI node.
+      // This should only happen in the unlikely event that we're syndicating
+      // paragraphs that do not exists on the target site.
+      $paragraphs_items[] = new ParagraphsItemEntity(array(
+        'field_name' => 'field_ding_news_paragraphs',
+        'bundle' => 'ding_paragraphs_text',
+        'field_ding_paragraphs_text' => array(
+          'und' => array(
+            array(
+              'value' => $bpi_properties['body'],
+              'format' => 'ding_wysiwyg',
+            ),
+          ),
+        ),
+      ));
+    }
+
+    return $paragraphs_items;
+  }
+
+  /**
+   * Descend (depth-first) into array with callback (by reference).
+   *
+   * This is basically a usable version of `array_walk_recursive` which visits
+   * all nodes (`array_walk_recursive` only visits leafs!)
+   *
+   * @param array $data
+   *   The data.
+   * @param callable $callback
+   *   The callback acception three parameters:
+   *   - $key: the current key.
+   *   - $value: the current value.
+   *   - $data: the full data object being traversed.
+   *
+   * @see https://www.php.net/manual/en/function.array-walk-recursive.php
+   */
+  private function traverse(array &$data, callable $callback) {
+    foreach ($data as $key => &$value) {
+      if (is_array($value)) {
+        $this->traverse($value, $callback);
+      }
+      $callback($key, $value, $data);
```

Comment:
@rimi-itk Take a look

Reply:
Only 1 in 4 uses of `traverse` collects values (in `replaceFilesWithUrls`), i.e. I don't think we should require the callback to return a value (or handle that it may not return anything). I've added a comment in the code to clarify that we collect values as a side effect.

Furthermore, I don't think getting rid of `traverse` (which I think you hint at) will make the code less complex. Without `traverse`, we'll have to come up with other recursive functions to run through the paragraphs structure and do whatever needs to be done, and the bodies of these functions will be just as complex as what we have now.

Reply:
Let's leave it at this.

I do not mind `traverse`. I would prefer methods without side effects - value collection as well as pass by reference. That should be achievable. I agree that the deeply nested paragraphs data structure does not make the situation easier.